### ISSUE TITLE
Worse test set perplexity than expected.

### ISSUE THREAD

**Issue by mrdrozdov** [urvashik/knnlm#4]
#### What is your question?

Using the provided snippet, I am only seeing 17.80 PPL on the test set of Wikitext-103 using faiss approx. distance, although the paper reports 16.50 PPL. Am I using the correct command?

#### Code

```
python eval_lm.py data-bin/wikitext-103 \
    --path wt103_checkpoint_best.pt \
    --sample-break-mode complete --max-tokens 3072 \
    --context-window 2560 --softmax-batch 1024 \
    --gen-subset test --dstore-filename dstore_train/dstore \
    --indexfile dstore_train/knn.index  \
    --model-overrides "{'knn_keytype': 'last_ffn_input'}" \
    --k 1024 --lmbda 0.25 --dstore-size 103225485 --knn-keytype last_ffn_input \
    --probe 32 --knnlm --knn-sim-func "do_not_recomp_l2" --no-load-keys
```

#### What have you tried?

I have rebuilt the knn index twice already, but I will try again.

#### What's your environment?

 - fairseq Version (e.g., 1.0 or master): `git+git@github.com:urvashik/knnlm.git@1c0a4e0ee29fc037b53a1449e7724af0e07dcc41#egg=fairseq`
 - PyTorch Version: 1.7.1
 - OS (e.g., Linux): Linux
 - How you installed fairseq: `pip install -e .`
 - Build command you used (if compiling from source): Followed instructions in README
 - Python version: 3.6
 - CUDA/cuDNN version: Titan X (Pascal) and CUDA 10.1 (from `nvcc -V`)
 - GPU models and configuration: n/a
 - Any other relevant information: n/a

**Comment by mrdrozdov** [comment]
The problem was my index job kept dying before completion (`build_dstore.py`). Interesting that it still gives some perplexity improvement even when partially complete.

After re-running `build_dstore.py` until completion, now I am getting the desired `16.46` perplexity with approximate distance.

**Comment by Sconcer** [comment]
Hi, I get worse ppl too (23.7), and I'm not sure where the bug is.
I'm not sure whether to use the trained index `checkpoints/knn.index.trained` in the final script to evaluate knnlm, or should I use `checkpoints/knn.index` for the `--indexfile`. Could you provide some instructions ? I want to make sure the `--indexfile` argument is correct.

**Comment by mrdrozdov** [comment]
It could be worth re-running the build dstore script, and make sure that it completed correctly. IIRC it takes multiple hours to run.

**Comment by araloak** [comment]
I believe I build the index right and i got two files, `knn.index.trained` and `knn.index` . I'm also confused about the index to be used. Which one did you use in knnlm evaluation?

**Comment by mrdrozdov** [comment]
The docs are pretty reliable, and clearly marks knn.index as the file to use. The .trained file is only used by the build_dstore script.

**Comment by araloak** [comment]
thanks, it helps a lot.
